extras/dht/dht.c

```diff
@@ -0,0 +1,141 @@
+ * 
+ *  Phase A - MCU pulls signal low for at least 18000 us
+ *  Phase B - MCU allows signal to float back up and waits 20-40us for DHT to pull it low
+ *  Phase C - DHT pulls signal low for ~80us
+ *  Phase D - DHT lets signal float back up for ~80us
+ *  
+ *  After this, the DHT transmits its first bit by holding the signal low for 50us
+ *  and then letting it float back high for a period of time that depends on the data bit.
+ *  duration_data_high is shorter than 50us for a logic '0' and longer than 50us for logic '1'.
+ *
+ *  There are a total of 40 data bits trasnmitted sequentially. These bits are read into a byte array
+ *  of length 5.  The first and third bytes are humidity (%) and temperature (C), respectively.  Bytes 2 and 4
+ *  are zero-filled and the fifth is a checksum such that:
+ *
+ *  byte_5 == (byte_1 + byte_2 + byte_3 + btye_4) & 0xFF
+ *
+*/
+
+/*  
+ *  @pin                    the selected GPIO pin
+ *  @interval               how frequently the pin state is checked in microseconds
+ *  @timeout                maximum length of time to wait for the expected pin state
+ *  @expected_pin_state     high (true) or low (false) pin state
+ *  @counter                pointer to external uint8_t for tallying the duration waited for the pin state
+*/
+
+bool dht_await_pin_state(uint8_t pin, uint8_t interval, uint8_t timeout, bool expected_pin_state, uint8_t * counter) {
+
+    for (*counter = 0; *counter < timeout; *counter+=interval) {
+        if (gpio_read(pin) == expected_pin_state) return true;
+        sdk_os_delay_us(interval);
+    }
+
+    return false;
+}
+
+/*  
+ *  
+ *  
+ *  @pin                    the selected GPIO pin
+ *  @internal_pullup        boolean to indicate whether the internal pullup resistor should be used
+ *  @humidity               pointer to external int8_t to store resulting humidity value
+ *  @temperature            pointer to external int8_t to store resulting temperature value
+ *  @status_code            pointer to external uint8_t to store status code corresponding to the following messages:
+ *                              0: "Success!"
+ *                              1: "Checksum failed, invalid data received from sensor..."
+ *                              2: "DHT did not initialize correctly, problem in phase 'B'..."
+ *                              3: "DHT did not initialize correctly, problem in phase 'C'..."
+ *                              4: "DHT did not initialize correctly, problem in phase 'D'..."
+*/
+
+bool dht_fetch_data(int8_t pin, bool internal_pullup, int8_t * humidity, int8_t * temperature, uint8_t * status_code) {
+    int8_t data[40] = {0};
+    int8_t result[5] = {0};
+    uint8_t i = 0;
+
+    uint8_t init_phase_duration = 0;
+    uint8_t duration_data_low = 0;
+    uint8_t duration_data_high = 0;
+
+    gpio_enable(pin, GPIO_OUT_OPEN_DRAIN);
+    gpio_set_pullup(pin, internal_pullup, false);

```

Comment:
Right..  `gpio_enable` is specifically coded not to mess with the pullup state, so `gpio_set_pullup` can be called once early on, and then no matter what's done later with `gpio_enable`, it will "do the right thing"..  We did it that way specifically because we've had the same issue in several other drivers as well, and it seemed silly for each driver to have to have its own code to deal with letting the user configure pullup state, so now they don't have to.